Modules/_elementtree.c

```diff
@@ -1632,46 +1632,59 @@ static PyObject *
 _elementtree_Element_remove_impl(ElementObject *self, PyObject *subelement)
 /*[clinic end generated code: output=38fe6c07d6d87d1f input=6133e1d05597d5ee]*/
 {
-    Py_ssize_t i;
-    int rc;
-    PyObject *found;
-
-    if (!self->extra) {
+    if (self->extra == NULL) {
         /* element has no children, so raise exception */
-        PyErr_SetString(
-            PyExc_ValueError,
-            "list.remove(x): x not in list"
-            );
-        return NULL;
+        goto not_found;
     }
 
-    for (i = 0; i < self->extra->length; i++) {
-        if (self->extra->children[i] == subelement)
+    Py_ssize_t i;
+    const Py_ssize_t length = self->extra->length;
+    for (i = 0; self->extra && i < self->extra->length; i++) {
+        if (self->extra->children[i] == subelement) {
             break;
-        rc = PyObject_RichCompareBool(self->extra->children[i], subelement, Py_EQ);
-        if (rc > 0)
+        }
+        PyObject *child = Py_NewRef(self->extra->children[i]);
+        int rc = PyObject_RichCompareBool(child, subelement, Py_EQ);
+        Py_DECREF(child);
+        if (self->extra ==  NULL || self->extra->length != length) {
+            goto fail;
+        }
+        if (rc > 0) {
             break;
-        if (rc < 0)
+        }
+        else if (rc < 0) {
             return NULL;
+        }
     }
 
-    if (i >= self->extra->length) {
+    // An extra check must be done if the mutation occurs at the very last
+    // step and removes or clears the 'extra' list (the condition on the
+    // length would not be satisfied any more).
+    if (self->extra == NULL || self->extra->length != length) {
+        goto fail;
+    }
+    else if (i >= self->extra->length) {
```

Comment:
Conditions `i >= self->extra->length` and `i < self->extra->length` are the same condition, and it is checked twice -- at the end of the loop and at the beginning of the loop. I think that the code can be simplified. Also, the check for `self->extra == NULL` is also performed multiple times (including the check at the beginning).

First try to rewrite the loop in the form:
```c
while (1) {
    if (...) goto ...;
    if (...) goto ...;
    ...
}
```
so every condition is only checked once. Then you can rewrite it in a loop with condition. The final code may even be simpler than the current code.

Reply:
I've managed to simplify the code.